### PR TITLE
Add a number of syntactic sugar features from libxml

### PR DESCRIPTION
**Pull request by scotchi**
When porting some [old code](https://github.com/directededge/directed-edge-bindings/commit/4be7b2bf7150e8b17f991168b9f76811961e77f6) from libxml-ruby to oga, there were a number of things that I found slightly nicer in the libxml API that were easy enough to pull over to Oga.  I realize it's probable that there were left out for a reason, but since the changes were minor, I thought I'd go ahead and throw them together to see if they'd be merged.

- Document#root_element

Gets the top level element in the document, rather than having to grab it from
the children.

- Node#<<

Makes it possible to append children or text in-line.

- Element#[]

Fetches an attribute's value using subscript operator

- Element#[]=

Sets an attribute's value using subscript operator

### CODE REVIEW

File: lib/oga/xml/element.rb [hunk 1]
```diff
@@ -91,6 +93,8 @@ def get(name)
         found ? found.value : nil
       end
 
+      alias [] get
+
```
Comment (YorickPeterse) [inline lib/oga/xml/element.rb]: Please use `alias_method` instead of just `alias` as `alias_method` is used in other places.

File: lib/oga/xml/element.rb [hunk 2]
```diff
@@ -64,6 +64,8 @@ def namespace_name=(name)
       #
       # @return [Oga::XML::Attribute]
       def attribute(name)
+        name = name.to_s if name.is_a?(Symbol)
+
```
Comment (YorickPeterse) [inline lib/oga/xml/element.rb]: These changes do not belong in this pull request. The method `split_name` already takes care of casting things to a String.

File: lib/oga/xml/element.rb [hunk 3]
```diff
@@ -113,6 +117,8 @@ def set(name, value)
         if found
           found.value = value
         else
+          name = name.to_s if name.is_a?(Symbol)
+
```
Comment (YorickPeterse) [inline lib/oga/xml/element.rb]: Same here. The code below should just use `split_name` similar to the `get` method, then this check is not needed.

File: lib/oga/xml/node.rb [hunk 4]
```diff
@@ -55,6 +55,21 @@ def children=(nodes)
...
+        child = Text.new(:text => child) if child.is_a?(String)
+
+        unless child.is_a?(Node)
+          raise ArgumentError, 'Can only add nodes or text.'
+        end
+
+        children << child
+        self
+      end
+
```
Comment (YorickPeterse) [inline lib/oga/xml/node.rb]: I do not want such a method in Oga. I think it's confusing that you're calling `Node#<<` / `Node#push` to push something into its _children_, which is an entirely different data structure.

Furthermore I'm not a fan of being able to pass both a `String` and a `Node` into `child`. I already made this mistake using `Node#replace` but I'd rather not introduce any more methods doing the same.

File: spec/oga/xml/document_spec.rb [hunk 5]
```diff
@@ -42,6 +42,27 @@
...
+  describe '#root_element' do
+    before do
+      @document = described_class.new
+    end
+
+    it 'returns first element' do
+      child = Oga::XML::Element.new(:name => 'foo')
+      @document.children << child
+      @document.root_element.should == child
+      @document.root_element.name.should == 'foo'
```
Comment (YorickPeterse) [inline spec/oga/xml/document_spec.rb]: This expectation is useless since the above expectation already checks for equality.

File: spec/oga/xml/document_spec.rb [hunk 6]
```diff
@@ -42,6 +42,27 @@
...
+    end
+
+    it 'returns first element even when other nodes present' do
+      @document.children << Oga::XML::Comment.new(:text => 'foo')
+      child = Oga::XML::Element.new(:name => 'foo')
+      @document.children << child
+      @document.root_element.should == child
+      @document.root_element.name.should == 'foo'
```
Comment (YorickPeterse) [inline spec/oga/xml/document_spec.rb]: Same here, just the equality check is more than enough.

File: spec/oga/xml/document_spec.rb [hunk 7]
```diff
@@ -42,6 +42,27 @@
     end
   end
 
+  describe '#root_element' do
+    before do
+      @document = described_class.new
+    end
+
+    it 'returns first element' do
```
Comment (YorickPeterse) [inline spec/oga/xml/document_spec.rb]: Please use `it 'returns the first element'`

File: spec/oga/xml/document_spec.rb [hunk 8]
```diff
@@ -42,6 +42,27 @@
...
+      child = Oga::XML::Element.new(:name => 'foo')
+      @document.children << child
+      @document.root_element.should == child
+      @document.root_element.name.should == 'foo'
+    end
+
+    it 'returns first element even when other nodes present' do
```
Comment (YorickPeterse) [inline spec/oga/xml/document_spec.rb]: Please use `it 'returns the first element even when it is preceded by another node'`

File: spec/oga/xml/element_spec.rb [hunk 9]
```diff
@@ -127,6 +127,10 @@
     it 'returns the value of an attribute' do
       @element.get('foo').should == 'bar'
     end
+
+    it 'returns the value of an attribute' do
```
Comment (YorickPeterse) [inline spec/oga/xml/element_spec.rb]: This spec should be scoped to a `describe '#[]'`.

File: spec/oga/xml/element_spec.rb [hunk 10]
```diff
@@ -161,6 +165,22 @@
...
+
+    it 'converts symbols to strings in names' do
+      @element[:foo] = 'foo'
+      @element[:foo].should == 'foo'
+      @element['foo'].should == 'foo'
+
+      @element['bar'] = 'bar'
+      @element[:bar].should == 'bar'
+      @element['bar'].should == 'bar'
+    end
```
Comment (YorickPeterse) [inline spec/oga/xml/element_spec.rb]: This spec should be scoped to a `describe '#[]='` outside of the `describe '#set'`.

Comment (YorickPeterse) [inline spec/oga/xml/element_spec.rb]: Also, this spec and the one for `Node#[]` can just check if `get` / `set` are called, they don't need to fully duplicate the tests just because they're aliases.

File: lib/oga/xml/document.rb [hunk 11]
```diff
@@ -57,6 +57,12 @@ def root_node
         self
       end
 
+      # Returns the top level element in the document.
+
```
Comment (YorickPeterse) [inline lib/oga/xml/document.rb]: Please remove the empty line here and adjust the documentation as follows:

```
# Returns the first Element in the document
#
# @return [Oga::XML::Element]
```

Right now the documentation is not entirely correct. While in most cases the first element will be the top-level element there's nothing preventing there from being multiple top-level elements (code wise). As such it's better to state that this method simply returns the first Element it finds.

Comment (scotchi) [inline lib/oga/xml/document.rb]: Is allowing multiple roots intentional?  Mightn't it be better to either throw an exception on adding a second root element (since that's not allowed by the XML spec)?

**Comment by scotchi** [comment]
I'll squash these all at the end, but for now I've handled the comments as separate commits for convenience.

File: lib/oga/xml/node.rb (hunk 4 above)
Comment (scotchi) [inline lib/oga/xml/node.rb]: So, first, I think there are two separate issues:

- Should this be allowed for nodes?
- Should this be allowed for nodes *and* text?

I'll start by saying that I could argue against both of the above, but after thinking about it a bit, have a slight preference for keeping at least one of them.

For appending nodes to a parent, I'll note that all of `REXML` `Nokogiri` and `libxml-ruby` *do* support appending nodes to parents via the `<<` operator.

In practice, I don't think it's ambiguous what the operator does.  Reading code from the other three libs doesn't cause any cognitive friction for me in this particular case.

The second question is I feel is harder -- my inclusion of text insertion wasn't so much out of a sense that the `<<` operator was the right solution, but that the current method felt clunky for such a common operation.

Since I wrote the code out to compare, I thought I'd include that here:

    require 'rexml/document'
    puts 'rexml:'
    
    doc = REXML::Document.new
    foo = REXML::Element.new('foo')
    bar = REXML::Element.new('bar')
    doc << foo
    foo << bar
    bar.text = 'baz'
    
    puts doc.to_s
    
    require 'nokogiri'
    puts 'nokogiri:'
    
    doc = Nokogiri::XML::Document.new
    foo = Nokogiri::XML::Element.new('foo', doc)
    bar = Nokogiri::XML::Element.new('bar', doc)
    doc << foo
    foo << bar
    bar << 'baz'
    
    puts doc.to_s
    
    require 'libxml'
    puts 'libxml:'
    
    doc = LibXML::XML::Document.new
    foo = LibXML::XML::Node.new('foo')
    bar = LibXML::XML::Node.new('bar')
    doc.root = foo
    foo << bar
    bar << 'baz'
    
    puts doc.to_s
    
    require 'oga'
    puts 'oga:'
    
    doc = Oga::XML::Document.new
    foo = Oga::XML::Element.new(:name => 'foo')
    bar = Oga::XML::Element.new(:name => 'bar')
    doc.children << foo
    foo.children << bar
    bar.inner_text = 'baz'
    
    puts doc.to_xml

File: lib/oga/xml/document.rb (hunk 11 above)
Comment (YorickPeterse) [inline lib/oga/xml/document.rb]: @scotchi This is intentional. HTML5 documents don't require the usual top-level elements, meaning you can have multiple root elements.

Come to think of it, because of that I'm not entirely sure if having this method makes a lot of sense since it wouldn't work very well for HTML5 documents.

Comment (scotchi) [inline lib/oga/xml/document.rb]: Hmm, I hadn't looked into the `HTML` module, but had assumed there was a parallel class tree there specialized for HTML-specific behavior.

To me it seems inevitable that some modality will be introduced (outside of just the parser) for the differences between XML and HTML.  It'd be simple enough for now to just set a field on the document from the parser, and then have methods like this do different things based on that field.  I get that that's not particularly elegant (but then, we're talking about parsing HTML), but leaving out behavior that everyone who's ever used an XML parser will expect because of HTML's quirks seems broken.

Comment (scotchi) [inline lib/oga/xml/document.rb]: In more concrete terms, this means for users of other libraries:

    doc << element

Has to be replaced by this:

    doc.children.last.children << element

Or, more verbose, but at least readable:

    root = doc.children.last
    root.children << element

Comment (YorickPeterse) [inline lib/oga/xml/document.rb]: @scotchi 

> Hmm, I hadn't looked into the HTML module, but had assumed there was a parallel class tree there specialized for HTML-specific behavior.

No, there are only some extra classes for the parser and for dealing with HTML entities. Classes such as `XML::Document` work for both XML and HTML documents. `XML::Document` _does_ have the `type` attribute which will return `:html` for HTML documents.

> To me it seems inevitable that some modality will be introduced (outside of just the parser) for the differences between XML and HTML.

This exists in a few places and mostly deals with serialisation/entity conversion.

>  I get that that's not particularly elegant (but then, we're talking about parsing HTML), but leaving out behavior that everyone who's ever used an XML parser will expect because of HTML's quirks seems broken.

Oga doesn't aim to be compatible with other parsers out there, instead it aims to be a more elegant and correct parser. I don't want to complicate Oga any further just because other parsers do so.

In the example of `root_element` it doesn't make sense to change behaviour between XML and HTML documents anyway. What would you have it return for HTML documents, all top-level elements? If so that directly conflicts with the name of the method and really isn't that useful.

> In more concrete terms, this means for users of other libraries:
> 
>     doc << element
> 
> Has to be replaced by this:
> 
>     doc.children.last.children << element
> 
> Or, more verbose, but at least readable:
> 
>     root = doc.children.last
>     root.children << element

These examples are confusing. If `Document#<<` would use `root_element` (which makes even less sense) then this would translate to `doc.children.first.children << element` and *not* `doc.children.last.children << element`.

Either way, to summarise:

* I'm not convinced `Node#<<`, `Document#<<` or something similar is useful as there already is a way of doing this by using `node_or_document.children << ...`
* `Document#root_element` doesn't strike me as very useful either (especially the more I think of it), and its behaviour doesn't work for certain HTML documents. Furthermore there's already `Document#root_node` (and `Node#root_node`) and the similar names may confuse users.

I feel that methods such as the above are better suitable for some oga-extensions Gem of sorts. If they become popular enough I may reconsider adding them to Oga.

File: lib/oga/xml/node.rb (hunk 4 above)
Comment (YorickPeterse) [inline lib/oga/xml/node.rb]: > For appending nodes to a parent, I'll note that all of REXML Nokogiri and libxml-ruby do support appending nodes to parents via the << operator.

As mentioned in another comment, implementing something in Oga just because other libraries do it is not a good reason. I want convincing use cases and I want to be certain additions are useful to all users.

> In practice, I don't think it's ambiguous what the operator does. Reading code from the other three libs doesn't cause any cognitive friction for me in this particular case.

This is subjective, and my opinion differs. I prefer the more explicit `some_node.children << foo` and I'm not convinced adding support for `some_node << foo` is worth the extra code, tests, maintenance burden, etc.

File: lib/oga/xml/document.rb (hunk 11 above)
Comment (scotchi) [inline lib/oga/xml/document.rb]: I'll stop pushing on `Element#<<`.  In that case, I agree that it's subjective, and as I stated at the beginning, I could argue either way on it.

However, I believe not having an accessor for the root element of the document is literally an API bug.  I don't mean simply all of the other 3 major Ruby XML implementations support that, I've literally never seen an XML API **in any language** that did not.

- Python: [xml.dom.Document.documentElement](https://docs.python.org/2/library/xml.dom.html#xml.dom.Document.documentElement)
- Java: [org.w3c.dom.Document.getDocumentElement](https://docs.oracle.com/javase/7/docs/api/org/w3c/dom/Document.html#getDocumentElement())
- PHP: [DOMDocument::$documentElement](http://php.net/manual/en/class.domdocument.php#domdocument.props.documentelement)
- Scala: [scala.xml.Document.docElem](http://www.scala-lang.org/api/2.10.4/index.html#scala.xml.Document@docElem:Node)

However, I'll roll that out to a separate pull request with some approximation of what I think the correct HTML and XML behavior would be.

Comment (YorickPeterse) [inline lib/oga/xml/document.rb]: @scotchi A lack of something offered by other libraries is not a bug, it's simply that: a lack of a feature. Having said that, let's move that to a separate pull request as you suggested.

**Comment by scotchi** [comment]
I squashed the commits now.  The only things remaining where the attribute aliases, so I reworked the commit message accordingly.

File: spec/oga/xml/element_spec.rb [hunk 12]
```diff
@@ -161,6 +168,14 @@
       @element.get('class').should == 'foo'
     end
 
+    it 'converts symbols to strings in names' do
```
Comment (YorickPeterse) [inline spec/oga/xml/element_spec.rb]: This is a bit picky but could you change this to `it 'supports the use of Symbols for attribute names'`?

Comment (scotchi) [inline spec/oga/xml/element_spec.rb]: Done.

**Comment by YorickPeterse** [comment]
Thanks!
